sway-core/src/debug_generation/dwarf.rs

```diff
@@ -0,0 +1,128 @@
+use std::fs::File;
+use std::os::unix::ffi::OsStringExt;
+use std::path::Path;
+
+use gimli::write::{
+    self, DebugLine, DebugLineStrOffsets, DebugStrOffsets, DwarfUnit, EndianVec, LineProgram,
+    LineString,
+};
+use gimli::{BigEndian, Encoding, LineEncoding};
+use sway_error::error::CompileError;
+use sway_types::Span;
+
+use crate::source_map::SourceMap;
+
+use object::write::Object;
+
+pub fn write_dwarf(
+    source_map: &SourceMap,
+    primary_dir: &Path,
+    primary_src: &Path,
+    out_file: &Path,
+) -> Result<(), CompileError> {
+    let encoding = gimli::Encoding {
+        format: gimli::Format::Dwarf64,
+        version: 5,
+        address_size: 8,
+    };
+
+    let program = build_line_number_program(encoding, primary_dir, primary_src, source_map)?;
+
+    program
+        .write(
+            &mut DebugLine::from(EndianVec::new(BigEndian)),
+            encoding,
+            &DebugLineStrOffsets::none(),
+            &DebugStrOffsets::none(),
+        )
+        .map_err(|err| {
+            sway_error::error::CompileError::InternalOwned(err.to_string(), Span::dummy())
+        })?;
+
+    let mut dwarf = DwarfUnit::new(encoding);
+    dwarf.unit.line_program = program;
+    // Write to new sections
+    let mut debug_sections = write::Sections::new(EndianVec::new(BigEndian));
+    dwarf
+        .write(&mut debug_sections)
+        .expect("Should write DWARF information");
+
+    let file = File::create(out_file).unwrap();
+    let mut obj = Object::new(
+        object::BinaryFormat::Elf,
+        object::Architecture::X86_64,
+        object::Endianness::Big,
+    );
+    debug_sections
+        .for_each(|section_id, data| {
+            let sec = obj.add_section(
+                [].into(),
+                section_id.name().into(),
+                object::SectionKind::Other,
+            );
+            obj.set_section_data(sec, data.clone().into_vec(), 8);
+            Ok::<(), ()>(())
+        })
+        .unwrap();
+
+    obj.write_stream(file).map_err(|err| {
+        sway_error::error::CompileError::InternalOwned(err.to_string(), Span::dummy())
+    })?;
+
+    Ok(())
+}
+
+fn build_line_number_program(
```

Comment:
It would be good to add tests for this function

Reply:
To effectively test this, we'll need external dwarf tools. My thought was that once we have code in the debugger to read back this data, we could add tests there. 

If you think we should just use external tools to test it now, I can try adding that (any suggestions on which test in the CI this will go under?). Any other suggestion on how to test it would be helpful too.

Reply:
I thought we might be able to add a test sway project, pass that path and dir into this function, and assert that the resulting LineProgram is translated into the correct JSON and/or dwarf string. Is that possible without adding additional dependencies?

Reply:
>Is that possible without adding additional dependencies?

I cannot see how. Because, this function only generates a DWARF object. It cannot read back that for us to compare. The reading part needs to be added into `forc-debug`. So when we add that, then we'll have an end-to-end flow and then it can be tested. Otherwise to read the DWARF object file, we'll need external tools such as `llvm-dwarf`.